WordPress/openverse
# Add codespell pre-commit hook

## Fixes



Fixes #378 by @dhruvkb 

## Description


This PR does the following:

1. Consolidates Python-based tool configuration into a single pyproject.toml. This file is auto-detected by all the relevant tools. Using it reduces the number of disparate files with varying naming conventions (dictated by the tools) at the repository root.
2. Adds and configures codespell, including additional documentation to assist with ongoing maintenance of the configuration.

Codespell seems fine enough and works pretty well. ~However, I would like to also explore using cspell. It has far greater flexibility in terms of the supported dictionaries and its configuration is significantly less obscure than Codespells. I'm going to leave this PR as a draft until I've tried cspell. If I decide to go with cspell, I'll open a new PR and keep the configuration consolidation I did for other Python-based tools in this PR. That seems useful regardless of which spell checker we choose.~ I tried out cspell and while it probably works great if you have a small project with little prose, it complains about every unknown word. At a glance I could not find anything in cspell's documentation about how to ignore unknown words, I suspect because its strategy for matching words is very different. Codespell is going to be much easier to maintain in the long run. It isn't perfect and will probably miss some things that cspell would catch, but it isn't worth the tremendous task of configuration and maintenance to use cspell. Keep in mind that "unknown words" includes most Vue related terms, all maintainer GitHub usernames (referenced frequently in project documentation), and so on and so on. In any case, Codespell is a good start. If we want to switch tools in the future we can do so. For now this gets us at least catching basic misspellings.

## Testing Instructions

Checkout the branch and run `just lint codespell`. Confirm there are no errors. Add a misspelling to a relevant file (one that isn't ignored) and confirm that codespell catches it when running `just lint codespell`. Also, review the list of ignored files and confirm that it's comprehensive and does not accidentally include files that _should_ be checked.

Look through the spelling changes and confirm that there are no problems with the fixes I've applied. There is a mix of automatically applied fixes that codespell generated as well as fixes I had to do manually (for when there were multiple suggestions for a misspelled word).

Finally, read through the configuration documentation and confirm that it makes sense, is clear, and can be used effectively to maintain the configuration moving forward.

## Checklist


- [N/A] I added or updated tests for the changes I made (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
